### PR TITLE
[public-api] Migrate envvarService

### PR DESCRIPTION
**Pull request by jeanp413**
## Description

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at 2d8b9cb</samp>

This pull request adds a new public API service for managing user environment variables for workspaces. It defines the service and its messages using Protobuf, generates client and server code for gRPC and Connect in Go and TypeScript, and implements the service handler and client in the server and dashboard components. It also refactors the dashboard component for environment variables to use the new public API service.

</details>

## Related Issue(s)

Fixes #

## How to test

- It should be able to CRUD env in `org`, `configuration` and `user` level
- Start workspace in `org`, envs in `org` + `user` should be in workspace
- Start workspace in `org` + `configuration`, envs in `org` + `configuration` + `user` should be in workspace
- Create a new `org`, there should only have `user` envs in workspace

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - jp-practical-guppy</li>
	<li><b>🔗 URL</b> - <a href="https://jp-practical-guppy.preview.gitpod-dev.com/workspaces" target="_blank">jp-practical-guppy.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - jp-practical-guppy-gha.20044</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-jp-practical-guppy%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-core-dev" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/public-api/gitpod/v1/envvar.proto
Comment (akosyakov) [inline components/public-api/gitpod/v1/envvar.proto:7]: @svenefftinge Could you review please?

File: components/public-api/gitpod/v1/envvar.proto
Comment (akosyakov) [inline components/public-api/gitpod/v1/envvar.proto:18]: Should we have `CreateEnvironmentVariable`?

It should also have oranization_id?

File: components/public-api/gitpod/v1/envvar.proto
Comment (akosyakov) [inline components/public-api/gitpod/v1/envvar.proto:45]: I think we should have here require `env_var_id`?

if env var belong to org then we should throw if envvar_repository_pattern is provided with invalid argument.

File: components/public-api/gitpod/v1/envvar.proto
Comment (akosyakov) [inline components/public-api/gitpod/v1/envvar.proto:55]: it should have only `env_var_id`

File: components/public-api/gitpod/v1/envvar.proto
Comment (akosyakov) [inline components/public-api/gitpod/v1/envvar.proto:31]: It is just `EnvironmentVariable` with oneof organization_id or repository_pattern?

File: components/public-api/gitpod/v1/envvar.proto
Comment (jeanp413) [reply components/public-api/gitpod/v1/envvar.proto:55]: I did that at first but then I would need to change dashboard logic too and thought maybe better to improve api as a follow up or should it be done at the same time?

File: components/public-api/gitpod/v1/envvar.proto
Comment (jeanp413) [reply components/public-api/gitpod/v1/envvar.proto:31]: there's also project en var

File: components/public-api/gitpod/v1/envvar.proto
Comment (akosyakov) [reply components/public-api/gitpod/v1/envvar.proto:55]: How big will be a change? We could do in 2 steps:
- change JSON-RPC and dashboard APIs
- then add shim

I have the same trouble now with https://github.com/gitpod-io/gitpod/pull/19062

File: components/public-api/gitpod/v1/envvar.proto
Comment (jeanp413) [reply components/public-api/gitpod/v1/envvar.proto:18]: why we need organization_id? old api doesn't have it, do you mean project_id?

Comment (jeanp413) [reply components/public-api/gitpod/v1/envvar.proto:18]: > Should we have CreateEnvironmentVariable?

Not sure current api does update or create at the same time, maybe as a follow up

Comment (akosyakov) [reply components/public-api/gitpod/v1/envvar.proto:18]: Sorry, I mean configuration_id. Project are named Configuration in gRPC Api.

File: components/public-api/gitpod/v1/envvar.proto
Comment (akosyakov) [reply components/public-api/gitpod/v1/envvar.proto:31]: Sorry I confused, I meant configuraiton_id or repository_pattern.

File: components/server/src/api/envvar-service-api.ts
Comment (jeanp413) [inline components/server/src/api/envvar-service-api.ts:164]: Not sure about thiss as there's no way to know if it's user or project env just with the id

File: components/server/src/api/envvar-service-api.ts
Comment (jeanp413) [inline components/server/src/api/envvar-service-api.ts:38]: Not sure about merging getWorkspaceEnvVars logic here too, the return value only has name and value, not id, is that fine? personally feels confusing to have many different return shapes for the same function

File: components/public-api/gitpod/v1/envvar.proto
Comment (akosyakov) [inline components/public-api/gitpod/v1/envvar.proto:32]: it is global type, i thin we should prefxi it then EnvrionmentVariableUserScope

or put them as nested, i.e.

```
message CreateEnvironmentVariableRequest {
  message UserScope {
...
```

File: components/public-api/gitpod/v1/envvar.proto
Comment (akosyakov) [inline components/public-api/gitpod/v1/envvar.proto:38]: @svenefftinge we could also think to have `EnvironmentVariableAdmission` enum instead, we use it as pattern for visibility in other places.

like
```proto
enum EnvironmentVariableAdmission {
  
  ENVIRONMENT_VARIABLE_ADMISSION_UNSPECIFIED;
  // only prebuild
  ENVIRONMENT_VARIABLE_ADMISSION_PREBUILD;
  // prebuild + workspace
  ENVIRONMENT_VARIABLE_ADMISSION_EVERYWHERE; // or   ENVIRONMENT_VARIABLE_ADMISSION_WORKSPACE;
}
```

File: components/public-api/gitpod/v1/envvar.proto
Comment (akosyakov) [inline components/public-api/gitpod/v1/envvar.proto:74]: Just `name` and `value`

File: components/public-api/gitpod/v1/envvar.proto
Comment (akosyakov) [inline components/public-api/gitpod/v1/envvar.proto:61]: just `name` and `value`

File: components/public-api/gitpod/v1/envvar.proto
Comment (akosyakov) [inline components/public-api/gitpod/v1/envvar.proto:65]: it breaks partial pattern here,

we should have then 
```
optional bool censored = 3; 
```

it would be a case for nested messages to override global, i..e
```proto
message UpdateEnvironmentVariableRequest {
  string env_var_id = 1; // required
  optional string name = 2;
  optional string value = 3;
  message UserScope {
    optional string repository_pattern = 1;
  }
  message ConfigurationScope {
    // configuration_id should not be here it is not updatable
    optional bool censored = 1;
  }
  oneof scope {
    optional UserScope userScope = 4;
    optional UserScope configurationScope = 5;
  }
}
```

File: components/public-api/gitpod/v1/envvar.proto
Comment (akosyakov) [reply components/public-api/gitpod/v1/envvar.proto:32]: given that they are not reusable for Update, i think it should be nested in CreateEnvironmentVariableRequest

Comment (jeanp413) [reply components/public-api/gitpod/v1/envvar.proto:32]: they are both used in update and create

Comment (akosyakov) [reply components/public-api/gitpod/v1/envvar.proto:32]: it will incorrect because Update should be partial, please see https://github.com/gitpod-io/gitpod/pull/19067#discussion_r1393986454 and https://www.notion.so/gitpod/Public-API-guideline-a010fa3d494544b78f28945e53f730ac?pvs=4#fe62cdc913df42babfbdc61b46044ed6

File: components/public-api/gitpod/v1/envvar.proto
Comment (svenefftinge) [inline components/public-api/gitpod/v1/envvar.proto:53]: can we add docs to whether these arguments are mandatory and how they are used?

File: components/public-api/gitpod/v1/envvar.proto
Comment (svenefftinge) [inline components/public-api/gitpod/v1/envvar.proto]: should an update not use the `env_var_id`?
Also do we want to allow updating the `name`?

Comment (svenefftinge) [reply components/public-api/gitpod/v1/envvar.proto]: Also is updating the scope something we need to allow? After all users can always delete and create

File: components/public-api/gitpod/v1/envvar.proto
Comment (svenefftinge) [reply components/public-api/gitpod/v1/envvar.proto:32]: we might not need and want to allow to update the scope of an env var

File: components/public-api/gitpod/v1/envvar.proto
Comment (jeanp413) [reply components/public-api/gitpod/v1/envvar.proto]: > should an update not use the env_var_id?

already fixed in latest commit

File: components/public-api/gitpod/v1/envvar.proto
Comment (jeanp413) [reply components/public-api/gitpod/v1/envvar.proto:53]: Not sure about this yet, see https://github.com/gitpod-io/gitpod/pull/19067#discussion_r1393701886

File: components/public-api/gitpod/v1/envvar.proto
Comment (svenefftinge) [reply components/public-api/gitpod/v1/envvar.proto:38]: Yes, that makes sense. we might have more scenarios in future

Comment (jeanp413) [reply components/public-api/gitpod/v1/envvar.proto:38]: How are these enum values to current `censored` value?
true === ENVIRONMENT_VARIABLE_ADMISSION_PREBUILD and false === ENVIRONMENT_VARIABLE_ADMISSION_EVERYWHERE?

Comment (akosyakov) [reply components/public-api/gitpod/v1/envvar.proto:38]: censored is ENVIRONMENT_VARIABLE_ADMISSION_PREBUILD

File: components/public-api/gitpod/v1/envvar.proto
Comment (jeanp413) [reply components/public-api/gitpod/v1/envvar.proto]: it's needed to actually call `listProjectEnvVars` internally, thoughts?

**Comment by jeanp413** [comment]
@akosyakov could you review again 🙏

File: components/public-api/gitpod/v1/envvar.proto
Comment (akosyakov) [inline components/public-api/gitpod/v1/envvar.proto:155]: generally looks good, not sure about it, maybe it is find to have it on global level?

**Comment by akosyakov** [comment]
@jeanp413 it looks ok, you can consider moving proto file and generated classes in another PR, to make this PR small, I will approve it to unblock.

**Comment by jeanp413** [comment]
@akosyakov moved the proto files to another PR https://github.com/gitpod-io/gitpod/pull/19085

**Comment by jeanp413** [comment]
Tested and it's working

**Comment by akosyakov** [comment]
@jeanp413 open for a review?

**Comment by jeanp413** [comment]
yep, need to rebase after the other PR is merged

**Comment by akosyakov** [comment]
@jeanp413 could you rebase and check out build failures? Is it still fine to test?

File: components/gitpod-protocol/src/public-api-converter.ts
Comment (akosyakov) [inline components/gitpod-protocol/src/public-api-converter.ts:238]: please add tests for it in the spec file

File: components/server/src/api/envvar-service-api.ts
Comment (akosyakov) [inline components/server/src/api/envvar-service-api.ts:63]: if it is uuid, please use uuid lib to validate, see how it is done in org service.

File: components/server/src/api/envvar-service-api.ts
Comment (akosyakov) [inline components/server/src/api/envvar-service-api.ts:69]: Why do we need to fetch everything again?  typorm supports partial updates, we should just partial change  to DB layer and add DB tests for env-var-service to check that it works, see for instance for update org settings

Comment (akosyakov) [reply components/server/src/api/envvar-service-api.ts:69]: i.e. what we discussed on teaming yesterday, use repo.update + repo.findOne(id)

File: components/server/src/api/envvar-service-api.ts
Comment (akosyakov) [inline components/server/src/api/envvar-service-api.ts:91]: it is not invalid arugment, but not found error code

File: components/server/src/api/envvar-service-api.ts
Comment (akosyakov) [inline components/server/src/api/envvar-service-api.ts:107]: As far as I understand typeorm has repo.save method which will create, update and return new object.

File: components/server/src/api/envvar-service-api.ts
Comment (akosyakov) [inline components/server/src/api/envvar-service-api.ts:133]: why is not enough to pass id?

**Review by akosyakov — COMMENTED** [review]
I left comment on gRPC implementation. I think at least we should fix error propagation, logic looks unnecessary complicated, but it seems because of underlying application service design. I wonder whether we can improve it, if it is too much work let's go with what we have for now.

File: components/dashboard/src/service/json-rpc-envvar-client.ts
Comment (akosyakov) [inline components/dashboard/src/service/json-rpc-envvar-client.ts:227]: ConnecError(unimplemented)

File: components/dashboard/src/data/projects/set-project-env-var-mutation.ts
Comment (akosyakov) [inline components/dashboard/src/data/projects/set-project-env-var-mutation.ts:15]: please update setup file to bump cache version and list env var classes to be safe

File: components/dashboard/src/service/json-rpc-envvar-client.ts
Comment (mustard-mh) [inline components/dashboard/src/service/json-rpc-envvar-client.ts:159]: Server has guard for this case, that if not found then create. So is it safe to update without check in shim? Or we want to make sure that it only do `update` but not `upsert`?

File: components/dashboard/src/data/projects/set-project-env-var-mutation.ts
Comment (mustard-mh) [reply components/dashboard/src/data/projects/set-project-env-var-mutation.ts:15]: code refer
- cache version [[1](https://github.com/gitpod-io/gitpod/blob/599e7b95c3522e53da2fa5bf8d65d643cfb02f61/components/dashboard/src/data/setup.tsx#L29)]
- classes [[2](https://github.com/gitpod-io/gitpod/blob/599e7b95c3522e53da2fa5bf8d65d643cfb02f61/components/dashboard/src/data/setup.tsx#L147)]

File: components/server/src/api/envvar-service-api.ts
Comment (mustard-mh) [inline components/server/src/api/envvar-service-api.ts:53]: we use `useCtxUid` (?) to get user id now (after you rebase to main), check other service-api files.

File: components/dashboard/src/service/json-rpc-envvar-client.ts
Comment (jeanp413) [reply components/dashboard/src/service/json-rpc-envvar-client.ts:159]: update should only update

File: components/server/src/api/envvar-service-api.ts
Comment (jeanp413) [reply components/server/src/api/envvar-service-api.ts:133]: it's needed for old api resource guard validation

Comment (jeanp413) [reply components/server/src/api/envvar-service-api.ts]: Already updated a couple 😅

**Comment by mustard-mh** [comment]
Add some nit changes.

❌  Failed to update user env, checked that list API has respond with id. Going to update code

image

File: components/gitpod-db/src/typeorm/user-db-impl.ts
Comment (jeanp413) [inline components/gitpod-db/src/typeorm/user-db-impl.ts:419]: @mustard-mh is it failing before or after this change?

Comment (mustard-mh) [reply components/gitpod-db/src/typeorm/user-db-impl.ts:419]: Error is throw by code above `An environment variable with this ID could not be found`. When I update user env value only

Comment (mustard-mh) [reply components/gitpod-db/src/typeorm/user-db-impl.ts:419]: We don't pass `envId` here

Comment (mustard-mh) [reply components/gitpod-db/src/typeorm/user-db-impl.ts:419]: Same on `envvar-service-api` side

Comment (jeanp413) [reply components/gitpod-db/src/typeorm/user-db-impl.ts:419]: 👍

Comment (jeanp413) [reply components/gitpod-db/src/typeorm/user-db-impl.ts:419]: Fixed

**Comment by mustard-mh** [comment]
I tried to create workspace with another ORG, it seems new workspace still includes previous ORG envs (the same repo)

**Comment by jeanp413** [comment]
you mean user env or project env?

**Comment by mustard-mh** [comment]
> org env? you mean user env or project env?

I mean org project env.

Seems something wrong in our code.

```log
mysql> select * from d_b_workspace where id='mustardmh-templatesvelt-5vcqsl66if7' \G;
*************************** 1. row ***************************
                   id: mustardmh-templatesvelt-5vcqsl66if7
         creationTime: 2023-11-20T13:58:03.696Z
              ownerId: f22cb39c-b108-4651-9212-c3bca4d391b6
           contextURL: var=value,var2=value2/https://github.com/mustard-mh/template-sveltejs
          description: mustard-mh/template-sveltejs - master
              context: *****
               config: *****
          imageSource: {"baseImageResolved":"docker.io/gitpod/workspace-full:latest"}
    imageNameResolved: eu.gcr.io/gitpod-core-dev/build/workspace-images:d1ce6d3571f4a5aca988999226fcac22f3b72326ab2becf9b0e80f561ed8395a
             archived: 0
            shareable: 0
        _lastModified: 2023-11-20 13:58:16.835081
              deleted: 0
                 type: regular
baseImageNameResolved: docker.io/gitpod/workspace-full@sha256:850314e6963b47c4e5d3ca18abe6345b435fa3423a735e7d94f93badf4502700
          softDeleted: NULL
               pinned: 0
      softDeletedTime: 
   contentDeletedTime: 
    basedOnPrebuildId: 
    basedOnSnapshotId: 
            projectId: 0c893085-6a88-4718-b5e9-4fddc52b5ae3
             cloneURL: https://github.com/mustard-mh/template-sveltejs.git
       organizationId: e6cfe015-326a-40d7-95a2-319f1994e051
1 row in set (0.28 sec)

ERROR: 
No query specified

mysql> select * from d_b_project_env_var \G;
*************************** 1. row ***************************
           id: 4c70516b-a522-4afe-91c9-7524c53bb5e6
    projectId: 0c893085-6a88-4718-b5e9-4fddc52b5ae3
         name: ORG_VAR_1
        value: *****
     censored: 1
 creationTime: 2023-11-20T13:26:27.858Z
      deleted: 0
_lastModified: 2023-11-20 13:26:27.860595
*************************** 2. row ***************************
           id: 9a6c52fa-d4d6-4259-9048-047894861b04
    projectId: 0c893085-6a88-4718-b5e9-4fddc52b5ae3
         name: ORG_VAR_2_VISIABLE
        value: *****
     censored: 0
 creationTime: 2023-11-20T13:27:08.348Z
      deleted: 0
_lastModified: 2023-11-20 13:27:08.352197
2 rows in set (0.48 sec)

ERROR: 
No query specified

mysql> select * from d_b_project \G;
*************************** 1. row ***************************
               id: 0c893085-6a88-4718-b5e9-4fddc52b5ae3
             name: template-sveltejs
         cloneUrl: https://github.com/mustard-mh/template-sveltejs.git
           teamId: d319b06f-3a73-4595-a341-ab63ff0fd65c
appInstallationId: 
     creationTime: 2023-11-20T13:26:04.970Z
    _lastModified: 2023-11-20 13:26:04.973654
    markedDeleted: 0
           userId: NULL
             slug: NULL
         settings: *****
1 row in set (0.42 sec)

ERROR: 
No query specified

mysql> 
```

**Comment by mustard-mh** [comment]
But not related to this PR's changes, tested all good.

**Comment by mustard-mh** [comment]
Let's wait preview env to validate updateUserEnv fixing. Once it's ok, we are good to go

**Comment by jeanp413** [comment]
> But ☝️ not relates to this PR's changes

😆

File: components/dashboard/src/service/json-rpc-envvar-client.ts
Comment (mustard-mh) [inline components/dashboard/src/service/json-rpc-envvar-client.ts:78]: I have one thing not so sure, that should `not found` use `Code.NotFound`? 

But it makes sense to use `Code.InvalidArgument` too since it shouldn't reach this function without UX.

Not blocker but want to align us all. cc @akosyakov @jeanp413

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (mustard-mh) [inline components/server/src/workspace/gitpod-server-impl.ts:1833]: If we update userEnv with `value` and `scope` changed, server will call `addUserEnvVar` even we have `envId` passed.

Not blocker as it's just like before 😌

File: components/dashboard/src/service/json-rpc-envvar-client.ts
Comment (jeanp413) [reply components/dashboard/src/service/json-rpc-envvar-client.ts:78]: should be not found, I remember changing it 🤔 , will fix it

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (jeanp413) [reply components/server/src/workspace/gitpod-server-impl.ts:1833]: yes

**Review by mustard-mh — APPROVED** [review]
Tested all good.

Bugs [[1](https://github.com/gitpod-io/gitpod/pull/19067#discussion_r1399271619)][[2](https://github.com/gitpod-io/gitpod/pull/19067#issuecomment-1819130113)] that not relates to this PR will create follow-up issues

File: components/dashboard/src/service/json-rpc-envvar-client.ts
Comment (akosyakov) [reply components/dashboard/src/service/json-rpc-envvar-client.ts:78]: please use ApplicationError as agreed today, drop ConnectError, only public-api-converter should translate them, no usage in any other places

Comment (akosyakov) [reply components/dashboard/src/service/json-rpc-envvar-client.ts:78]: we clean up today other places: https://github.com/gitpod-io/gitpod/pull/19092

Comment (mustard-mh) [reply components/dashboard/src/service/json-rpc-envvar-client.ts:78]: @akosyakov  Maybe rebase on your PR then do it together here?

Comment (akosyakov) [reply components/dashboard/src/service/json-rpc-envvar-client.ts]: Not sure what you mean, mentioned PR already merged.

It is rather easy to fix, open json and grpc impl files, do
```
search: ConnectError\((".*"), Code.InvalidArgument\);
replace: ApplicationError(ErrorCodes.BAD_REQUEST, $1);
```

then do the same for NOT_FOUND

**Pull request merged**
